design/design-doc-template-stacks-experience.md

```diff
@@ -0,0 +1,505 @@
+      # This is a top-level object to namespace hook configurations.
+      # There can be hooks for multiple different types of events.
+      hooks:
+        # Post create is triggered after an instance is created
+        postCreate:
+          # These are the templates which should be rendered when an object
+          # of the type above is seen.
+          #
+          # Defaults for the variables can be defined in the default values
+          # for the CRD fields, using the standard Kubernetes mechanism for
+          # specifying CRD field default values.
+          # Note that this is a list of objects, so multiple can be
+          # specified.
+          - directory: wordpress
+        # Post create is triggered after an instance is changed
+        postUpdate:
+          - directory: wordpress
+EOF
+```
+
+The `stack.yaml` should be in the build root of the repository. For most
+single-stack repositories, this means the root of the repository.
+
+For some realistic samples, see the [helm engine
+example][helm-engine-example] or the [quick start
+example][quick-start-example] in the template stack experience
+repository.
+
+### Specifying default values
+
+Default values will be specified in the CRD itself, using the [standard
+mechanism][kubernetes-default-values] for specifying default values for
+CRD fields. Here is an example excerpt from a realistic CRD, where the
+CRD's `spec.image` field is configured with a default:
+
+```yaml
+kind: CustomResourceDefinition
+metadata:
+  creationTimestamp: null
+  name: wordpressinstances.wordpress.samples.stacks.crossplane.io
+spec:
+  group: wordpress.samples.stacks.crossplane.io
+  names:
+    kind: WordpressInstance
+    plural: wordpressinstances
+  scope: ""
+  validation:
+    openAPIV3Schema:
+      description: WordpressInstance is the Schema for the wordpressinstances API
+      properties:
+        ...
+        spec:
+          type: object
+          properties:
+            ...
+            image:
+              type: string
+              description: A custom wordpress container image id to use
+              # Defaults are specified like this, using the schema validation for CRD fields.
+              # For more about how this works with CRDs, see:
+              # https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#defaulting
+              default: "wordpress:4.6.1-apache"
+```
+
+For the full example, see the sample in the
+[template stack experience repository](https://github.com/suskin/template-stack-experience/tree/master/wordpress-workload/go-templates/default-variables).
+
+### How processing objects works under the covers
+
+The imagined implementation for how objects sent to the stack for
+rendering become resources is that the object's fields become the input
+for the template rendered by the controller. For more details, see the
+[detailed examples](https://github.com/suskin/template-stack-experience)
```

Comment:
this is another example where it hurts readability and cognitive processing of these concepts by having the examples in another repo and just linking to the root of it, putting the burden on the reader to navigate through that other repo and discover on their own the concepts being mentioned here.

i think this design would be easier to parse and understand if 1 example was put inline here that demonstrates a template consuming input from the CR.

Reply:
Thanks for pointing this out! This was originally intended to be a bit hand-wavey; since we are talking about supporting multiple engines, it was unclear whether getting into specifics of how things are transformed made sense yet. Maybe it makes sense now though, since we have a "complete" example in the repo.

Hmm, this link should probably just link to the quick start now.

I'll look into whether there's a nice concise example that can be pasted in here or linked to directly. I'm thinking the helm version of the app stack in the quick start could work well.